<launch>
  <arg name="scope" default="false" />
  
  <!-- Controller parameters -->
  <rosparam file="$(find quad_control)/config/drone_param.yaml" />
  <rosparam file="$(find quad_control)/config/controller_param.yaml" />  

  <env name="GAZEBO_MODEL_PATH" value="${GAZEBO_MODEL_PATH}:$(find rotors_gazebo)/models"/>
  <env name="GAZEBO_RESOURCE_PATH" value="${GAZEBO_RESOURCE_PATH}:$(find rotors_gazebo)/models"/>
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find quad_control)/worlds/arena_with_columns.world" />
    <arg name="paused" value="true" />
    <arg name="use_sim_time" default="true"/>
  </include>

  <group ns="hummingbird">
    <include file="$(find rotors_gazebo)/launch/spawn_mav.launch">
      <arg name="mav_name" value="hummingbird" />
      <arg name="model" value="$(find rotors_description)/urdf/mav_with_vi_sensor.gazebo" />
      <arg name="log_file" value="hummingbird"/>
    </include>

    <node name="odometry_node" type="odometry_node" pkg="quad_control">
      <remap  from="/odometry"      to="/hummingbird/ground_truth/odometryNED"/>
      <remap  from="/p"             to="/hummingbird/odometryNED/p"/>
      <remap  from="/eta"           to="/hummingbird/odometryNED/eta"/>
    </node> 

    <node name="trajectory_node" type="trajectory_node" pkg="quad_control">
      <remap  from="/gazebo"        to="/gazebo/model_states"/>
      <remap  from="/p"             to="/hummingbird/ground_truth/odometryNED"/>
      <remap  from="/traj/p"        to="/hummingbird/trajectory/p"/>
      <remap  from="/traj/yaw"      to="/hummingbird/trajectory/yaw"/>
    </node>

    <node name="outer_loop_node" type="outer_loop_node" pkg="quad_control">
      <remap  from="/traj_topic"    to="/hummingbird/trajectory/p"/>
      <remap  from="/odom_topic"    to="/hummingbird/odometryNED/p"/>
      <remap  from="/est_topic"     to="/hummingbird/estimator"/>
      <remap  from="/mu_hat"        to="/hummingbird/mu_hat"/>
    </node>

    <node name="eta_reference_node" type="eta_reference_node" pkg="quad_control">
      <remap  from="/sub_topic"     to="/hummingbird/mu_hat"/>
      <remap  from="/traj_topic"    to="/hummingbird/trajectory/yaw"/>
      <remap  from="/eta_ref_topic" to="/hummingbird/eta_reference"/>
    </node>

    <node name="inner_loop_node" type="inner_loop_node" pkg="quad_control">
      <remap  from="/eta_ref_topic" to="/hummingbird/eta_reference"/>
      <remap  from="/odom_topic"    to="/hummingbird/odometryNED/eta"/>
      <remap  from="/mu_topic"      to="/hummingbird/mu_hat" />
      <remap  from="/est_topic"     to="/hummingbird/estimator" />
      <remap  from="/pub_topic"     to="/hummingbird/command/wrenchNED"/>
    </node>

    <node name="estimator_node" type="estimator_node" pkg="quad_control">
      <remap  from="/eta"           to="/hummingbird/odometryNED/eta" />
      <remap  from="/p"             to="/hummingbird/odometryNED/p"/>
      <remap  from="/cmd"           to="/hummingbird/command/wrenchNED"/>
      <remap  from="/est_topic"     to="/hummingbird/estimator"/>
    </node>

    <!--
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" />

    <node pkg="tf" type="static_transform_publisher" name="maptf" args="0 0 0 0 0 0  world /map  10"/>
    <node pkg="tf" type="static_transform_publisher" name="wNED" args="0 0 0 1.57079 0 3.14159  world worldNED  10"/>
    <node pkg="tf" type="static_transform_publisher" name="bodyNED" args="0 0 0 1.57079 0 3.14159  hummingbird/base_link hummingbird/base_linkNED  10"/>
    
    
     <node name="rviz" pkg="rviz" type="rviz" args="-d $(find quad_control)/config/drone.rviz"  />
     -->
  </group>

  <node name="plotjuggler" pkg="plotjuggler" type="plotjuggler" 
      args="--layout $(find quad_control)/config/plotJugglerConfig.xml" 
      if="$(arg scope)"/>
</launch>
